Revive the xNVMe Python language bindings

safl opened the pull request:
xNVMe previously had Python bindings, however, these where manually written wrappers which were not properly tested, thus they became stale and were removed from the repository.

However, recently when providing examples of how to dynamically load xNVMe from C and Python, then it seemed worth-while to explore just how much is possible via ``ctypes``. With a couple of non-intrusive changes to the public C API (adding a couple of includes), then it is possible to auto-generate a complete ctypes-wrapping for the C API. With the essentials auto-generated, then using ``ctypes`` is a breeze.

With this PR a full xNVMe C API mapping is available along with automatic dynamic loading. This module can serve for those wanting to experiment with the API from the comforts of Python. Additionally, as a reference for adding Cython bindings as well, for those going beyond poking at the API.

- [x] Add testcases and testplan for CIJOE invoking and verifying the library
- [x] Add testing to CI
- [x] Add description of the Python module in the docs

Have a look at the ``toolbox/py_xnvme_pkg/README.rst`` for more information.

Baekalfen commented on .github/workflows/verify.yml:
```diff
@@ -370,6 +370,13 @@ jobs:
     - name: Execute 'xnvme library-info'
       run: xnvme library-info
 
+    - name: Build and install Python bindings
+      run: |
+        pip3 install pytest
+        cd toolbox/py_xnvme_pkg
+        make
+        make test
+
```
Just a heads-up. You've only added it to `build-macos`. Linux is at line 326.

safl replied:
Right. Fixed!
Windows and FreeBSD are still untested.
I have made a note of that in the TODO/Known Issues in ``docs/python/index.rst``.

Baekalfen commented on toolbox/py_xnvme_pkg/setup.py:
```diff
@@ -0,0 +1,43 @@
+
+
+setuptools.setup(
+    name="py-xNVMe",
```
This is the name for `pip uninstall [name]`

safl replied:
Right, fixing that.

Baekalfen commented on toolbox/py_xnvme_pkg/Makefile:
```diff
@@ -0,0 +1,123 @@
+generate-ctypes:
+	@echo "## py: gen"
+	@clang2py ${XNVME_CAPI_HEADERS} > xnvme/ctypes_mapping.py
+	@echo "## Remember to also call 'make format-all' to fix format"
```
`format-all ` should probably be `patch-ctypes`

safl replied:
Right! Thanks!

Baekalfen commented on toolbox/py_xnvme_pkg/Makefile:
```diff
@@ -0,0 +1,123 @@
+TOOLBOX_DIR?=../
+PROJECT_NAME=xnvme
+XNVME_CAPI_HEADERS=$(shell ls ../../include/libxnvme* | grep -v util | grep -v libxnvmec)
+XNVME_SO_PATH=$(shell pkg-config xnvme --variable=libdir)/libxnvme-shared.so
```
This is Linux specific. I would also recommend, that we look for the lib in the local repo. For example on Mac: `$GIT_ROOT/builddir/lib/libxnvme-shared.dylib`

safl replied:
Good spot. Usage of this is completely removed, it now relies on the library being installed on the system as that is where ``clang2py`` looked for headers anyway, so the library and the headers are needed there regardless.

Baekalfen commented on toolbox/py_xnvme_pkg/Makefile:
```diff
@@ -0,0 +1,123 @@
+.PHONY: uninstall
+uninstall:
+	@echo "## py: make uninstall"
+	@pip3 uninstall ${PROJECT_NAME} --yes || echo "Cannot uninstall => That is OK"
```
`setup.py` dictates the name to provide here. It's currently `py-xNVMe`, and not `xnvme` as `${PROJECT_NAME}` resolves to.

safl replied:
Fixed.

Baekalfen commented on toolbox/py_xnvme_pkg/Makefile:
```diff
@@ -0,0 +1,123 @@
+.PHONY: generate-ctypes
+generate-ctypes:
+	@echo "## py: gen"
+	@clang2py ${XNVME_CAPI_HEADERS} > xnvme/ctypes_mapping.py
```
I had to specify additional arguments for macOS: `@clang2py --clang-args="-I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk/usr/include/ -I../../include/" ${XNVME_CAPI_HEADERS} > xnvme/ctypes_mapping.py`

safl replied:
Hmm, curious. This is probably something to explore further. To get this moving along then the build-testing is restricted to a single platform, the package generated and then tested on all the usual suspects.

Baekalfen commented on toolbox/py_xnvme_pkg/xnvme/ctypes_mapping.py:
```diff
@@ -0,0 +1,4272 @@
+# -*- coding: utf-8 -*-
```
I would recommend not having the auto-generated mapping in the repo. As @safl noted offline, the ordering of the output is not stable, so git-diffs are impossible to check. I would recommend just adding it as part of the build process, and publish it as part of a package on PyPi.

safl replied:
I agree. They are have been removed now.
The facilitate the build then a ci-job now builds the the sdist package and added as an artifact.
By doing so, then we get to do two things: 1) Automate the build of the package. 2) Test that the runtime-dependencies are met on all the usual targets.

Baekalfen commented on toolbox/py_xnvme_pkg/xnvme/library_loader.py:
```diff
@@ -0,0 +1,50 @@
+            )
+    except subprocess.CalledProcessError:
+        pass
+
```
I would recommend adding a third-priority path with the local build: `yield f"{git_root}/builddir/lib/libxnvme-shared.dylib"`

safl replied:
I don't like having a dependency on the source-repository in the library loader. Libraries should be installed such that ``pkg-config`` and regular tooling can locate them.
What I have done instead is add a platform specific extension, such that it loads using ``.so``, ``.dll`` and ``.dylib`` depending the platform it is running on.

Baekalfen commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+  # Alpine
+  sudo apk add clang-libs
+
+  # macOS
+  sudo brew ??
```
Installed with Xcode. You wouldn't get as far as building xNVMe without it.

Baekalfen commented on python/Makefile:
```diff
@@ -0,0 +1,122 @@
+# invoke: 'make uninstall', 'make install'
+endef
+.PHONY: default
+default: uninstall install
```
I think we should make the default `generate-ctypes patch-ctypes`, as I understand you'd idiomatically write `make && make install`.

Baekalfen commented on python/Makefile:
```diff
@@ -0,0 +1,122 @@
+.PHONY: test
+test:
+	@echo "## py: make test"
+	@python3 -m pytest --pyargs xnvme
```
Omit `--pyargs`

safl replied:
As stated elsewhere, it is there for a reason.

Baekalfen commented on python/bin/xpy_dev_open:
```diff
@@ -0,0 +1,51 @@
+
+import xnvme.ctypes_mapping as capi
+
+capi.guard_unloadable()
```
I think it's better to move this to the library.

safl replied:
How about this would go on the list of things to refine in ``docs/python/index.rst``? Have done so now.

Baekalfen commented on python/bin/xpy_enumerate:
```diff
@@ -0,0 +1,36 @@
+
+import xnvme.ctypes_mapping as capi
+
+capi.guard_unloadable()
```
Same here

safl replied:
How about this would go on the list of things to refine in ``docs/python/index.rst``?

Baekalfen commented on python/setup.py:
```diff
@@ -0,0 +1,43 @@
+    package_dir={"": "./"},
+    packages=setuptools.find_packages(where="./"),
+    python_requires=">=3.6",
+    install_requires="pytest",
```
Install wouldn't require `pytest`. Either use `tests_require` or now that you have a `python/pyproject.toml`, put it there instead.

safl replied:
I have gone with the approach of installing the testcases, this allows the ease-of-use of running the pytest via the package-discovery. This is why pytest is added as a runtime requirement.

Baekalfen replied:
I still think we should not have it here, and put it in either `tests_require` or `pyproject.toml`, as it's not part of normal runtime usage.

Also, isn't it supposed to be a list? `["pytest"]`?

Baekalfen replied:
The reason not to do it, would be because it's not a runtime dependency. I think it would be exceedingly rare somebody else is running our test suite. And as soon as we merge my Cython PR, the tests would no longer succeed without proper configuration.

But why not put it under the test requirements, if it's for testing?

safl replied:
Adding to that, this is why ``python3 -m pytest --pyargs xnvme`` is added as a bold note in the docs right after installation.

safl replied:
Also, this does not dictate that all tests are installed with the package, just the ones useful for the use-case described above.

Baekalfen replied:
How about a different approach? Instead of implementing these as a pytest, why not just a simple function?

```py
def self_test():
     # Self-test for the dynamically loaded library

     # The 'capi' is loaded upon module access, this checks that it succeeded
     assert capi.is_loaded, "C API isn't loaded"

     # This verifies that the loaded library actually is the xNVMe library, by checking for
     # the existance of a xNVMe specific funtion.
     assert capi.xnvme_enumerate, "xnvme_enumerate is not present in library. Library isn't loaded correctly"

     # Verifies that at at least one function is callable
     assert capi.xnvme_libconf_pr(capi.XNVME_PR_DEF), "Failed to call function 'xnvme_libconf_pr in xNVMe. Library isn't loaded correctly"

     print('xNVMe ctypes bindings loaded correctly!')
```

Then in the docs: `python3 -c 'import xnvme.ctypes_bindings as xnvme; xnvme.self_test()'`

safl replied:
That is also an approach, the down-side is that then the tests need to added in multiple places and not behaving as nicely as ``pytest``.

I do like the messages on the asserts, so I have added that to the tests.

I don't see any reason not to use ``pytest`` for this, the only counter-argument is that it is "unusual" but it is no more unusual than ``pytest`` describing the use-case in their documentation and providing test-collection for it. So, its a well-documented and supported feature of ``pytest``.

Should the addition of ``pytest`` cause us grief, then it is no worse than we can remove it and replace it with something like the approach above. However, for now I see no reason to, it only makes things simpler/easier to use.

Baekalfen replied:
> the only counter-argument is that it is "unusual"

The other argument is that this solution is short-lived. As soon as we merge my Cython-to-Python bindings, it won't work anymore.

> Should the addition of pytest cause us grief, then it is no worse than we can remove it and replace it with something like the approach above.

If you want to, then keep and let's see.

Baekalfen commented on python/requirements.txt:
```diff
@@ -0,0 +1,5 @@
+black
+ctypeslib2
+# All
+pytest
```
See comment in setup.py

safl replied:
done.

Baekalfen commented on python/xnvme/library_loader.py:
```diff
@@ -0,0 +1,60 @@
+import platform
+import subprocess
+
+SHARED_EXT = {
+    "linux": "so",
+    "windows": "dll",
+    "darwin": "dylib",
+}
```
You can probably get Python's `sysconfig` to report this

safl replied:
Ohh, neat, ``SHLIB_SUFFIX`` looks like something that is useful for that purpose. Changing it to use that instead.

safl replied:
``sysconfig.get_config_var("SHLIB_SUFFIX")`` seems like the thing to use?

Baekalfen replied:
I haven't investigated it, but it (incorrectly?) provides `.so` on macOS
```
>>> import sysconfig
>>> sysconfig.get_config_var("SHLIB_SUFFIX")
'.so'
```

safl replied:
No very useful then. Reverting back.

safl replied:
Yup. reverted.

safl commented on .github/workflows/verify.yml:
```diff
@@ -249,11 +249,62 @@ jobs:
+      run: |
+        make config-slim
+        make
+        make install
```
Right. removed. This also removes the testing done at this stage which is also done in the later ``build-linux`` for Alpine,

safl commented on .github/workflows/verify.yml:
```diff
@@ -249,11 +249,62 @@ jobs:
+        make generate-ctypes patch-ctypes install build-sdist test
+        for file in dist/*.tar.gz; do mv $file ${file//xnvme/xnvme-py}; done
+
+    - name: Upload Python package
+      uses: actions/upload-artifact@v2
+      with:
+        name: python-pkg
+        path: python/dist/xnvme-py-*.tar.gz
```
This is just the creation of a GitHUB artifact, such that the package is available for the other jobs.

safl commented on .github/workflows/verify.yml:
```diff
@@ -381,8 +456,21 @@ jobs:
+    - name: Install and test xNVMe Python bindings
+      run: |
+        python3 -m pip install xnvme-py*.tar.gz --user
+        python3 -m pytest --pyargs xnvme
```
The ``-pyargs`` tells it to search the xnvme module installed on the system. By doing this, then we do not need to known where it is located.

Baekalfen replied:
You're right. I misunderstood the docs. Cool trick

safl commented:
@Baekalfen I have addressed your feedback, some with changes, some with clarification, and one thing post-poned as "nice-to-have" / added to ``docs/python/index.rst`` for future fixing.

safl commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+the :ref:`sec-python-development` section.
+
+.. note:: The **xNVMe** Python package comes with pytests, thus, you can verify
+   your installation by running: ``python3 -m pytest --pyargs xnvme``.
```
Yes, tests are shipped with the package.
``--pyargs xnvme`` is there to collect tests from the package.

safl commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+ctypes
+======
+
+Python ctypes_ comes with Python, that is, with the reference implementation of
```
That is Sphinx-doc notation for producing a link.

safl commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+======
+
+Python ctypes_ comes with Python, that is, with the reference implementation of
+the Python intepreter. When using the ``ctypes`` bindings, not other runtime
```
fixed.

safl commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+Usage examples
+--------------
+
+See the files in ``bin/*``, at the time of writing this, there were three
+examples, two of which require no arguments, these are run like so::
```
Fixed.

safl commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+Development
+===========
+
+When generating the ctypes_ bindings, then **xNVMe** must be installed on the
+system, otherwise ``clang`` fails to locate the header-files. The bindings are
```
Fixed.

safl commented on docs/python/index.rst:
```diff
@@ -0,0 +1,167 @@
+
+Before installing, then ensure that the following runtime dependencies are met:
+
+* Python 3.6+
```
Yeah, but that is because it only builds on Alpine which has some Python 3.7+ version, however, there are several instances of testing it on Python 3.6 (Leap / Ubuntu), also the ``setup.py`` has ``python_requires=">=3.6"``.

safl commented:
@Baekalfen found some of your comments on the commits. Addressed those as well and pushed. Thanks for all the input! Your review improved this PR in a lot of ways:

* Reduced the ``build-python`` from ~50 seconds to  ~30 seconds
* Removed the generator-requirement needing the xNVMe libraries installed on the system
* Removed generated bindings from the repos
* Improved the ``Makefile`` helper, it now behaves idiomatically
* Several fixes on the docs and additions to todo

Thank you!

Baekalfen commented on python/bin/xpy_dev_open:
```diff
@@ -0,0 +1,51 @@
+import argparse
+import sys
+
+import xnvme.ctypes_mapping as capi
```
Not a hard requirement, but maybe consider `import xnvme.ctypes_mapping as xnvme`. Doesn't make much difference here, but in a larger program, it's maybe a bit ambiguous to call it `capi`.

safl replied:
Sure. Changed to ``xnvme``.

safl commented:
Thanks @Baekalfen !
